internal/runbits/checkout/checkout.go

```diff
@@ -68,35 +73,38 @@ func (r *Checkout) Run(ns *project.Namespaced, branchName, cachePath, targetPath
 	}
 
 	var branch *mono_models.Branch
-	if branchName == "" {
-		branch, err = model.DefaultBranchForProject(pj)
-		if err != nil {
-			return "", errs.Wrap(err, "Could not grab branch for project")
+	commitID := ns.CommitID
+
+	switch {
+	// Fetch the branch the given commitID is on.
+	case commitID != nil:
+		for _, b := range pj.Branches {
+			if belongs, err := model.CommitBelongsToBranch(ns.Owner, ns.Project, b.Label, *commitID, r.auth); err == nil && belongs {
```

Comment:
It's unfortunate the number of API calls we're going to have to make in order to determine this. I did some poking around the VCS API and didn't see anything. It _may_ be worth asking if something exists for this use case, but I doubt it so I'll leave it up to you.